Push preliminary AWS deployment documentation (#467)
2i2c-org/pilot-hubs: docs/howto/operate/add-aws-hub.md

# Add a new hub in a AWS kops-based cluster

The idea behind this guide is showcase the process of setting up a 
[kops](https://kops.sigs.k8s.io/getting_started/aws/)-based AWS
cluster and manually deploy a new hub on top of it using our deployer tool.
This is a preliminary but fully functional and manual process. Once
[#381](https://github.com/2i2c-org/pilot-hubs/issues/381) is resolved, we should be able
to automate the hub deployment process as we currently do with the GKE-based hubs.

```{note}
We will continue working toward a definitive one once we figured out some of the
discussions outlined in [#431](https://github.com/2i2c-org/pilot-hubs/issues/431).
```

## Pre-requisites

1. Follow the instructions outlined in
[Set up and use the the deployment scripts locally](./manual-deploy.md) to set up the
local environment and prepare `sops` to encrypt and decrypt files.

2. Install the awscli tool (you can use pip or conda to install it in the environment)

> and configure it to use the provided AWS user credentials. Follow this
> [guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config)
> for a quick configuration process.

Reviewer: ```suggestion
and configure it to use the provided AWS user credentials. [Follow this guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config)
for a quick configuration process.
```

```{note}
The customer with AWS admin privileges should have created a user for you with full
privileges. We will probably explore fine-graining the permissions actually needed
in the short-term.
```

3. Export some helpful AWS environment variables (because "aws configure" doesn't export
these environments vars for kops to use, so we do it manually).

```bash
export AWS_PROFILE=<cluster_name>
export AWS_ACCESS_KEY_ID=$(aws configure get aws_access_key_id)
export AWS_SECRET_ACCESS_KEY=$(aws configure get aws_secret_access_key)
```

## Create an AWS kops-based cluster

1. From the root directory on this repo, `cd` into the `kops` directory
2. Set up a *state* bucket for kops. This bucket will store the cluster
[state](https://kops.sigs.k8s.io/state/).

``` bash

> export KOPS_STATE_STORE=s3://<2i2c>-<hub-name>-kops-state

Reviewer 2: What does <2i2c> mean here? Should it be `2i2c-<hub-name>-kops-state`?

Author: You are totally right, thanks for noticing it!

> export KOPS_STATE_STORE=s3://<2i2c>-<hub-name>-kops-state

Reviewer: ```suggestion
export KOPS_STATE_STORE=s3://2i2c-<hub-name>-kops-state
```

aws s3 mb $KOPS_STATE_STORE --region <region>
```

```{note}

> Kops [recommends](https://kops.sigs.k8s.io/getting_started/aws/#cluster-state-storage)
> versioning the S3 bucket in case you ever need to revert or recover a previous state

Reviewer: ```suggestion
Kops [recommends versioning the S3 bucket](https://kops.sigs.k8s.io/getting_started/aws/#cluster-state-storage) in case you ever need to revert or recover a previous state
```

store.
```

### Create and render a kops config file

You can use 
[one](https://github.com/2i2c-org/pilot-hubs/blob/master/kops/farallon.jsonnet) of the
existing [jsonnet](https://jsonnet.org/) specifications as a "template" for your cluster.
You may need to tweak zones, names and instances, the rest is boilerplate to create a
kops-based cluster accordingly to some specification already outlined in 
[#28](https://github.com/2i2c-org/pangeo-hubs/issues/28). Once you have your jsonnet

> specification ready, you need to render it to create the config file kops understand.

Reviewer: I'm not sure what "understand" means in this context

Author: Kops is able to process the incoming config file to create the desired cluster: https://kops.sigs.k8s.io/cli/kops_create/

Author: Alternative suggestion accepted :wink: .

Author: Resolving the conversation as per 👍

1. Render the `kops` config file with

```bash
jsonnet <cluster_name>.jsonnet -y > <cluster_name>.kops.yaml
```

> 2. Regrettably, the rendering creates yaml file with with 3 dots at the end, you can 

Reviewer: Is this file just an unnecessary extra created by the step that has no purpose? Could you add a short sentence noting why this file is OK to be deleted?

Author: I am not being clear enough, I think... the generated yaml file will be used with the `kops create` command... the issue here is that jsonnet creates this file with 3 points at the very end of the file.

Author: Resolving since I think the text is clear enough :wink:

delete it with

```bash
sed -i '' -e '$ d' <cluster_name>.kops.yaml
```

```{note}
In Linux you will need to use instead: `sed -i '$ d' <cluster_name>.kops.yaml`
```

### Create the cluster

1. Create the cluster configuration and push it to s3 with

```bash
kops create -f <cluster_name>.kops.yaml
```

2. You will need to a ssh key pair before actually creating the cluster

```bash
ssh-keygen -f ssh-key
mv ssh-key <cluster_name>.key
mv ssh-key.pub <cluster_name>.key.pub
```

```{note}

> You will need to relocate and encrypt the private key before pushing your changes to the

Reviewer 3: Add commands for this too?

Reviewer: Or we could add a separate section on encrypting private keys, since that is probably a generic thing and not AWS-specific, and cross-link it?

Author: Both options are OK... I think I would probably add the step here for now (I was thinking in that direction before pushing this one) and then we can add a new section and cross-reference in another PR...

Author: Added a new 4h step to relocate and encrypt the private key.

repository.
```

> 3. Build the cluster with (notice that you are passing the ssh public key you just 

Reviewer: ```suggestion
3. Build the cluster with the following command (notice that you are passing the ssh public key you just 
```

created in the previous step)

```bash
kops update cluster <cluster_name>hub.k8s.local --yes --ssh-public-key <cluster_name>.key.pub --admin
```

```{note}
The `--admin` at the end will modify your `~/.kube/config` file to point to the new 
cluster.
```

If everything went as expected, the cluster will be created after some minutes, and you 
should be able to validate it with

```bash

> kops validate cluster --wait 10m

Author: > Should we just always run the coredns patch first before validating?

Yep, that also crossed through my head before and you validated it with your comment :wink:

> kops validate cluster --wait 10m

Author: OK, asked for validation after applying the patches.

```

But validation will not pass until this next section is done.

### Apply workaround to run CoreDNS on the master node

> More details on this [issue](https://github.com/kubernetes/kops/issues/11199)

Reviewer: ````suggestion
```{seealso}
More details [in this issue](https://github.com/kubernetes/kops/issues/11199)
```
````

Author: Nice, thanks!

> 1. After the failed validation finished (you can run these command in another terminal 

Reviewer: ```suggestion
1. After the failed validation finished (you can run these commands in another terminal 
```

if you are impatient ;-) patch CoreDNS with

```bash
kubectl -n kube-system patch deployment kube-dns --type json --patch '[{"op": "add", "path": "/spec/template/spec/tolerations", "value": [{"key": "node-role.kubernetes.io/master", "effect": "NoSchedule"}]}]'
deployment.apps/kube-dns patched

kubectl -n kube-system patch deployment kube-dns-autoscaler --type json --patch '[{"op": "add", "path": "/spec/template/spec/tolerations", "value": [{"key": "node-role.kubernetes.io/master", "effect": "NoSchedule"}]}]'
```

### Create an EFS for your cluster

> 1. Install boto3 with pip or conda

Reviewer 3: ```suggestion
1. Install `boto3` with pip or conda
```

Author: Thanks!

2. Create an [EFS](https://aws.amazon.com/efs/) file system for this hub with

```bash

> python3 setup-efs.py <cluster_name>hub.k8s.local us-east-2

Reviewer 3: ```suggestion
python3 setup-efs.py <cluster_name>hub.k8s.local <region-of-cluster>
```

Author: Thanks!

```

This will output an fs-<xxxxxxxx> id. You should use that value 
(it should be something like `fs-<id>.efs.<region>.amazonaws.com`) in
the `basehub.nfsPVC.nfs.serverIP` at you hub config file. 

## Deploy the new hub

1. First, `cd` back to the root of the repository

2. Generate a kubeconfig that will be used by the deployer with

```bash
KUBECONFIG=secrets/<cluster_name>.yaml kops export kubecfg --admin=730h <cluster_name>hub.k8s.local
```

3. Encrypt (in-place) the generated kubeconfig with sops

```bash
sops -i -e secrets/<cluster_name>.yaml
```

4. Generate a new config file for your cluster

> You can use 
> [one](https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/farallon.cluster.yaml)
> of the existing cluster config files as a "template" for your cluster.
> You may need to tweak names, `serverIP` and singleuser's images references. Make sure 

Reviewer: ```suggestion
You can use of the existing cluster config files as a "template" for your cluster (for example, [here is the Farallon Institute config file](https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/farallon.cluster.yaml)).
You may need to tweak names, `serverIP` and singleuser's images references. Make sure 
```

This is a preliminary but fully functional and manual process. Once
[#381](https://github.com/2i2c-org/pilot-hubs/issues/381) is resolved, we should be able
to automate the hub deployment process as we currently do with the GKE-based hubs.

```{note}
We will continue working toward a definitive one once we figured out some of the
discussions outlined in [#431](https://github.com/2i2c-org/pilot-hubs/issues/431).
```

## Pre-requisites

1. Follow the instructions outlined in
[Set up and use the the deployment scripts locally](./manual-deploy.md) to set up the
local environment and prepare `sops` to encrypt and decrypt files.

2. Install the awscli tool (you can use pip or conda to install it in the environment)
and configure it to use the provided AWS user credentials. [Follow this guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config)
for a quick configuration process.

```{note}
The customer with AWS admin privileges should have created a user for you with full
privileges. We will probably explore fine-graining the permissions actually needed
in the short-term.
```

> 3. Export some helpful AWS environment variables (because "aws configure" doesn't export

Reviewer 3: ```suggestion
3. Export some helpful AWS environment variables (because `aws configure` doesn't export
```

these environments vars for kops to use, so we do it manually).

```bash
export AWS_PROFILE=<cluster_name>

> export AWS_ACCESS_KEY_ID=$(aws configure get aws_access_key_id)

Reviewer 3: I think an easier / safer way is to put these in `~/.aws/credentials` and just export `AWS_PROFILE`. Does `aws configure` populate `~/.aws/credentials`?

Author: > Does aws configure populate ~/.aws/credentials?

Yep, it does. In fact, those are already in the credentials file (this is why you can get them with `aws configure get`).
But AFAIK, when you export the profile name, the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY env variables are not exported as well and kops seems to need those accordingly to the docs (check the last box at that section): https://kops.sigs.k8s.io/getting_started/aws/#setup-iam-user.

Author: Since this is minor, I will not block the merge on it.
We can always iterate and change it later down the road.
Btw, feel free to continue the discussion.

export AWS_SECRET_ACCESS_KEY=$(aws configure get aws_secret_access_key)
```

## Create an AWS kops-based cluster

1. From the root directory on this repo, `cd` into the `kops` directory
2. Set up a *state* bucket for kops. This bucket will store the cluster
[state](https://kops.sigs.k8s.io/state/).

``` bash

> export KOPS_STATE_STORE=s3://2i2c-<hub-name>-kops-state

Reviewer 3: ```suggestion
export KOPS_STATE_STORE=s3://2i2c-<cluster-name>-kops-state
```

Where `cluster-name` also matches what's under `config/hubs/<cluster-name>.cluster.yaml`?

Author: Yep, it should be cluster instead of hub!

> aws s3 mb $KOPS_STATE_STORE --region <region>

Reviewer 3: ```suggestion
aws s3 mb $KOPS_STATE_STORE --region <region>
aws s3api put-bucket-versioning --bucket $KOPS_STATE_STORE --versioning-configuration Status=Enabled
```

I think this should enable versioning?

Author: Yep, I agree. That should be the command to enabled versioning.

You can use 
of the existing [jsonnet](https://jsonnet.org/) specifications as a "template" for your cluster (for example, [here is the Farallon Institute specification](https://github.com/2i2c-org/pilot-hubs/blob/master/kops/farallon.jsonnet)).
You may need to tweak zones, names and instances, the rest is boilerplate to create a
kops-based cluster accordingly to some specification already outlined in 
[#28](https://github.com/2i2c-org/pangeo-hubs/issues/28). Once you have your jsonnet
specification ready, you need to render it to create the config file kops understand.

1. Render the `kops` config file with

```bash
jsonnet <cluster_name>.jsonnet -y > <cluster_name>.kops.yaml
```

2. Regrettably, the rendering creates yaml file with with 3 dots at the end, you can 
delete it with

```bash
sed -i '' -e '$ d' <cluster_name>.kops.yaml
```

```{note}
In Linux you will need to use instead: `sed -i '$ d' <cluster_name>.kops.yaml`
```

### Create the cluster

1. Create the cluster configuration and push it to s3 with

```bash
kops create -f <cluster_name>.kops.yaml
```

2. You will need to a ssh key pair before actually creating the cluster

```bash

> ssh-keygen -f ssh-key

Reviewer 3: ```suggestion
ssh-keygen -f <cluster-name>.key
```

This automatically creates the two, so you don't need the `mv`

mv ssh-key <cluster_name>.key
mv ssh-key.pub <cluster_name>.key.pub
```

3. Build the cluster with the following command (notice that you are passing the ssh public key you just 
created in the previous step)

```bash
kops update cluster <cluster_name>hub.k8s.local --yes --ssh-public-key <cluster_name>.key.pub --admin
```

```{note}
The `--admin` at the end will modify your `~/.kube/config` file to point to the new 
cluster.
```

If everything went as expected, the cluster will be created after some minutes.

```{note}
If you try to validate the cluster, validation will not pass until this next section is done.
```

4. Relocate and encrypt the generated keys

After creating the cluster, you will need to relocate the public and private keys with

```bash
mv <cluster_name>.key ssh-keys/<cluster_name>.key

> mv <cluster_name>.key.pub ssh-keys/<cluster_name>.key.pub

Reviewer 3: What do you think about not checking in the public key? We can always regenerate it from the private key with `ssh-keygen -y -f <cluster-name>.key`

Author: I agree, I will remove that line and make a note about how to recover it.
Btw, we should probably remove the public keys we have already pushed into the repo.

```bash
KUBECONFIG=secrets/<cluster_name>.yaml kops export kubecfg --admin=730h <cluster_name>hub.k8s.local
```

3. Encrypt (in-place) the generated kubeconfig with sops

```bash
sops -i -e secrets/<cluster_name>.yaml
```

4. Generate a new config file for your cluster

You can use of the existing cluster config files as a "template" for your cluster (for example, [here is the Farallon Institute config file](https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/farallon.cluster.yaml)).
You may need to tweak names, `serverIP` and singleuser's images references. Make sure 
you set up the `profileList` section to be compatible with your kops cluster (ie. match 
the `node_selector` with the proper `instance-type`).

5. Set `proxy.https.enabled` to `false`. This creates the hubs without trying to give 
them HTTPS, so we can appropriately create DNS entries for them.

6. Deploy the hub (or hubs in case you are deploying more than one) without running the
test with

```bash
python3 deployer deploy <cluster_name> --skip-hub-health-test
```

7. Get the AWS external IP for your hub with (supposing your hub is `staging`):

```bash
kubectl -n staging get svc proxy-public
```

> Create a CNAME record for `staging.foo.2i2c.cloud` and point it to the AWS external IP.

Reviewer 3: This needs to be repeated for staging and prod as well, no?

Author: Yes, I will make a note about that as well.
